File: chesskit-server/src/algorithms/estimateElo.ts
import { ceilsNumber } from "@/lib/math";
import { EstimatedElo, PositionEval } from "@/types/eval";

/**
 * Оценивает примерный рейтинг по среднему centipawn loss (CPL) игры.
 * Берёт среднее CPL для белых/чёрных и преобразует в рейтинг, используя формулы Chesskit.
 */
export const computeEstimatedElo = (
  positions: PositionEval[],
  whiteElo?: number,
  blackElo?: number
): EstimatedElo | undefined => {
  if (positions.length < 2) {
    return undefined;
  }

  const { whiteCpl, blackCpl } = getPlayersAverageCpl(positions);

  const whiteEstimatedElo = getEloFromRatingAndCpl(
    whiteCpl,
    whiteElo ?? blackElo
  );
  const blackEstimatedElo = getEloFromRatingAndCpl(
    blackCpl,
    blackElo ?? whiteElo
  );

  return { white: whiteEstimatedElo, black: blackEstimatedElo };
};

// Получаем оценку позиции в центимату по логике Chesskit
const getPositionCp = (position: PositionEval): number => {
  const line0 = position?.lines?.[0];
  if (!line0) {
    // Явно сигнализируем, что движок не вернул линию для позиции
    throw new Error("missing_primary_line");
  }

  if (line0.cp !== undefined) {
    return ceilsNumber(line0.cp, -1000, 1000);
  }

  if (line0.mate !== undefined) {
    // Используем «бесконечность» и клиппим до [-1000;1000], как в Chesskit
    return ceilsNumber(line0.mate * Infinity, -1000, 1000);
  }

  throw new Error("missing_cp_and_mate");
};


/**
 * Для каждой стороны суммирует потери (cpl) и делит на фиксированное количество ходов.
 * Не пропускает ходы — все позиции должны иметь cp или mate.
 */
const getPlayersAverageCpl = (
  positions: PositionEval[]
): { whiteCpl: number; blackCpl: number } => {
  // Начинаем со значения cp первой позиции
  let previousCp = getPositionCp(positions[0]);

  // Суммируем потери cp для каждой стороны. Не пропускаем ходы,
  // даже если оценок нет, поскольку getPositionCp бросит исключение
  const { whiteCpl, blackCpl } = positions.slice(1).reduce(
    (acc, position, index) => {
      const cp = getPositionCp(position);

      if (index % 2 === 0) {
        // белые ходили
        acc.whiteCpl += cp > previousCp ? 0 : Math.min(previousCp - cp, 1000);
      } else {
        // чёрные ходили
        acc.blackCpl += cp < previousCp ? 0 : Math.min(cp - previousCp, 1000);
      }

      previousCp = cp;
      return acc;
    },
    { whiteCpl: 0, blackCpl: 0 }
  );

  // Делим на фиксированное количество ходов для каждой стороны
  return {
    whiteCpl: whiteCpl / Math.ceil((positions.length - 1) / 2),
    blackCpl: blackCpl / Math.floor((positions.length - 1) / 2),
  };
};

// Lichess formula: перевод среднего CPL в рейтинг ELO
const getEloFromAverageCpl = (averageCpl: number) =>
  3100 * Math.exp(-0.01 * averageCpl);

// Обратная функция: ожидаемое CPL для заданного рейтинга
const getAverageCplFromElo = (elo: number) =>
  -100 * Math.log(Math.min(elo, 3100) / 3100);

/**
 * Комбинирует CPL и известный рейтинг игрока.
 * Если рейтинг известен, корректируем оценку в сторону игрока.
 */
const getEloFromRatingAndCpl = (
  gameCpl: number,
  rating: number | undefined
): number => {
  const eloFromCpl = getEloFromAverageCpl(gameCpl);
  if (!rating) return eloFromCpl;

  const expectedCpl = getAverageCplFromElo(rating);
  const cplDiff = gameCpl - expectedCpl;
  if (cplDiff === 0) return eloFromCpl;

  if (cplDiff > 0) {
    return rating * Math.exp(-0.005 * cplDiff);
  } else {
    return rating / Math.exp(-0.005 * -cplDiff);
  }
};

File: app/src/main/java/com/example/chessanalysis/analysis/EstimateElo.kt
package com.example.chessanalysis.analysis

import com.example.chessanalysis.PositionEval
import kotlin.math.ceil
import kotlin.math.exp
import kotlin.math.floor
import kotlin.math.ln
import kotlin.math.min

data class EstimatedEloResult(
    val white: Int?,
    val black: Int?
)

object EstimateElo {

    fun computeEstimatedElo(
        positions: List<PositionEval>,
        whiteElo: Int?,
        blackElo: Int?
    ): EstimatedEloResult {
        if (positions.size < 2) {
            return EstimatedEloResult(null, null)
        }

        val (whiteCpl, blackCpl) = getPlayersAverageCpl(positions)

        val whiteEstimatedElo = getEloFromRatingAndCpl(whiteCpl, whiteElo ?: blackElo)
        val blackEstimatedElo = getEloFromRatingAndCpl(blackCpl, blackElo ?: whiteElo)

        return EstimatedEloResult(
            white = whiteEstimatedElo?.toInt(),
            black = blackEstimatedElo?.toInt()
        )
    }

    private fun getPositionCp(position: PositionEval): Int {
        val line = position.lines.firstOrNull()
            ?: throw IllegalStateException("No lines in position")

        return when {
            line.cp != null -> MathUtils.ceilsNumber(line.cp.toDouble(), -1000.0, 1000.0).toInt()
            line.mate != null -> {
                val value = line.mate * Double.POSITIVE_INFINITY
                MathUtils.ceilsNumber(value, -1000.0, 1000.0).toInt()
            }
            else -> throw IllegalStateException("No cp or mate in line")
        }
    }

    private fun getPlayersAverageCpl(positions: List<PositionEval>): Pair<Double, Double> {
        var previousCp = getPositionCp(positions[0])

        var whiteCpl = 0.0
        var blackCpl = 0.0

        positions.drop(1).forEachIndexed { index, position ->
            val cp = getPositionCp(position)

            if (index % 2 == 0) {
                whiteCpl += if (cp > previousCp) 0.0 else min((previousCp - cp).toDouble(), 1000.0)
            } else {
                blackCpl += if (cp < previousCp) 0.0 else min((cp - previousCp).toDouble(), 1000.0)
            }

            previousCp = cp
        }

        val whiteAvg = whiteCpl / ceil((positions.size - 1) / 2.0)
        val blackAvg = blackCpl / floor((positions.size - 1) / 2.0)

        return Pair(whiteAvg, blackAvg)
    }

    // Source: https://lichess.org/forum/general-chess-discussion/how-to-estimate-your-elo-for-a-game-using-acpl-and-what-it-realistically-means
    private fun getEloFromAverageCpl(averageCpl: Double): Double {
        return 3100.0 * exp(-0.01 * averageCpl)
    }

    private fun getAverageCplFromElo(elo: Double): Double {
        return -100.0 * ln(min(elo, 3100.0) / 3100.0)
    }

    private fun getEloFromRatingAndCpl(gameCpl: Double, rating: Int?): Double? {
        val eloFromCpl = getEloFromAverageCpl(gameCpl)
        if (rating == null) return eloFromCpl

        val expectedCpl = getAverageCplFromElo(rating.toDouble())
        val cplDiff = gameCpl - expectedCpl

        if (cplDiff == 0.0) return eloFromCpl

        return if (cplDiff > 0) {
            rating * exp(-0.005 * cplDiff)
        } else {
            rating / exp(-0.005 * -cplDiff)
        }
    }
}